vignettes/data-repository.Rmd

```diff
@@ -0,0 +1,236 @@
+
+The data file (".tsv") and the metadata file (".yml") are automatically staged after writing. `write_delim_git()` returns the hashes of both the data file and the metadata file. The names of the vector contain the path of the files, relative to the root of the _git_ repository.
+
+```{r}
+status(repo)
+write_delim_git(x = x, file = "my_data", repo = repo, sorting = c("y", "x"))
+status(repo)
+```
+
+Overwriting data which has different metadata than the original data will throw an error because this can potentially lead to large diffs. Differences in metadata occur when:
+
+- the sorting changes
+- variables are added, removed or renamed
+- the class of a variable changes
+- the levels of a factor changes
+
+However you can bypass this by using `override = TRUE`. This will completely ignore the existing metadata and create new metadata based on the new `data.frame`.
+
+```{r error=TRUE}
+y <- x
+y$extra <- x$x
+write_delim_git(y, file = "my_data", repo = repo, sorting = c("y", "x"))
+write_delim_git(
+    y, file = "my_data", repo = repo, sorting = c("y", "x"), 
+    override = TRUE
+)
+```
+
+### Reading data
+
+Retrieving data is straight forward. Use `read_delim_git` and provide the `file` and the `repo`. The retrieved data is identical to the original data after applying the ordering of variables and observations.
+
+```{r}
+y_stored <- read_delim_git(file = "my_data", repo = repo)
+str(y_stored)
+y_sorted <- y[order(y$y, y$x), colnames(y_stored)]
+rownames(y_sorted) <- NULL
+stopifnot(all.equal(y_sorted, y_stored))
+```
+
+### Committing changes
+
+Commit the changes using `commit()`
+
+```{r}
+write_delim_git(x, file = "sub/test", repo = repo, sorting = "x")
+status(repo)
+add(repo, ".")
+status(repo)
+commit(repo, message = "Initial commit")
+status(repo)
+```
+
+### Removing data
+
+Data objects are files, so you remove them with `rm_file`. 
+
+```{r}
+# remove a single file
+rm_file(repo = repo, path = "my_data.tsv")
+rm_file(repo = repo, path = "my_data.yml")
+status(repo)
+# undo the remove by resetting to the last commit
+reset(commits(repo)[[1]], "hard")
+status(repo)
+# remove all data objects in a given path
+rm_data(repo = repo, path = ".", type = "both")
+status(repo)
+```
+
+Sometimes you want to write a bunch of data objects to the repository and you cannot guarantee that all data objects will remain after the update. The workflow below is intended for such case.
+
+1. use `rm_data()` with `type = "tsv"` to remove **all** ".tsv" in a specific path of the repository
+1. use `write_delim_git()` to write all data objects. Since the ".yml" are still present, any existing metadata will be used.
+1. use `type = "yml"` in `rm_data()` to remove **any** ".yml" which have no associated ".tsv"
+1. `commit()` the changes
+
+```{r}
+# undo the remove by resetting to the last commit
+reset(commits(repo)[[1]], "hard")
```

Comment:
is this line necessary here?

Reply:
The alternative is rewritting the data objects. resetting the repo to the last commit is more elegant.